server/src/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -1920,12 +1920,13 @@ private Volume orchestrateDetachVolumeFromVM(long vmId, long volumeId) {
             // Mark the volume as detached
             _volsDao.detachVolume(volume.getId());
 
-            // volume.getPoolId() should be null if the VM we are detaching the disk from has never been started before
-            DataStore dataStore = volume.getPoolId() != null ? dataStoreMgr.getDataStore(volume.getPoolId(), DataStoreRole.Primary) : null;
-
-            volService.revokeAccess(volFactory.getVolume(volume.getId()), host, dataStore);
-
-            handleTargetsForVMware(hostId, volumePool.getHostAddress(), volumePool.getPort(), volume.get_iScsiName());
+            // volumePool() should be null if the VM we are detaching the disk from has never been started before
+            // only revoke access on volumes that are actually on a datastore
+            if (volumePool != null) {
```

Comment:
Should the original logic be retained? What happens if volumePool is null?

Reply:
If the volumePool was null initially the dataStore was null and revokeAccess did nothing. Also I don't think it should be null as the ```handleTargetsForVMWare(hostId,volumePool.getHostAddress()...``` would crash.